text/0000-custom-dst.md

```diff
@@ -0,0 +1,279 @@
+- Feature Name: custom\_dst
+- Start Date: 02 March, 2016
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow Rust code to define custom fat pointers, and define slice code in terms of
+these new operators (instead of the current `std::raw::Repr`).
+
+# Motivation
+[motivation]: #motivation
+
+One major missing feature of Rust is custom dynamically sized types; it's
+impossible to implement a slice type in Rust, for example. This has led to
+things like the `std::raw::Repr` trait, a stopgap measure. It's also very
+difficult to represent things like views into 2d slices; for example,
+pixel buffers. In many places, something other than the Rust slice type is
+needed (if you have both a width and a height, or if you have a width, height,
+and stride, or even further).
+
+# Detailed design
+[design]: #detailed-design
+
+The most important addition to the language shall be one new trait:
+
+```rust
+#[lang = "dst"]
+unsafe trait Dst {
+    type Meta: Copy;
+    fn size_of_val(&self) -> usize;
+}
+```
+
+The `Dst` trait should be unsafe, because it's very easy to give an incorrect
+implementation of `size_of_val`, and safe and unsafe code *must* rely on it
+being correct to not create undefined behavior.
+
+Note that it is specified that, if `size_of::<<T as Dst>::Meta>() == 0`, then
+`size_of::<*const T>()` == `size_of::<*const ()>`. This is really important for
+C compatibility, because our current compatibility with flexible arrays is
+really terrible.
+
+To actually get use of the improved fat pointers, we must add four new
+intrinsics (the names are up for bikeshedding):
+
+```rust
+// core
+mod ptr {
+    pub use intrinsics::{make_fat_ptr, make_fat_ptr_mut};
+    pub fn fat_ptr_meta<T: Dst + ?Sized>(ptr: &T) -> T::Meta {
+        unsafe {
+            core::intrinsics::fat_ptr_meta(ptr)
+        }
+    }
+}
+mod mem {
+    pub fn size_of_prelude<T: ?Sized>() -> usize {
+        unsafe {
+            core::intrinsics::size_of_prelude::<T>()
+        }
+    }
+}
+mod intrinsics {
+    extern "rust-intrinsic" {
+        fn size_of_prelude<T: ?Sized>() -> usize;
+        fn fat_ptr_meta<T: ?Sized + Dst>(ptr: *const T) -> T::Meta;
+        fn make_fat_ptr<T: Dst + ?Sized>(data: *const (),
+            meta: U::::Meta) -> *const T;
+        fn make_fat_ptr_mut<T: Dst + ?Sized>(data: *mut (),
```

Comment:
Yeah, you probably could. It's all design. I personally just thought it felt more "rusty".